File: frontend/lib/models/draft_analysis.dart
import 'package:flutter/material.dart';
import 'package:messageai/models/situation_type.dart';
import 'package:messageai/models/response_template.dart';

/// Model for draft message analysis results
/// Extends tone analysis with confidence scoring and suggestions
class DraftAnalysis {
  // Tone analysis fields
  final String tone;
  final String? intensity;
  final String? urgencyLevel;
  final String? intent;
  final Map<String, dynamic>? contextFlags;
  final String? reasoning;
  
  // Draft-specific fields
  final int confidenceScore; // 0-100
  final AppropriatenessLevel appropriateness;
  final List<String> suggestions;
  final List<String> warnings;
  final List<String> strengths;
  
  // ✅ NEW: Phase 2 fields
  final SituationDetection? situationDetection;
  final List<ResponseTemplate>? suggestedTemplates;

  const DraftAnalysis({
    required this.tone,
    this.intensity,
    this.urgencyLevel,
    this.intent,
    this.contextFlags,
    this.reasoning,
    required this.confidenceScore,
    required this.appropriateness,
    required this.suggestions,
    required this.warnings,
    required this.strengths,
    // NEW
    this.situationDetection,
    this.suggestedTemplates,
  });

  factory DraftAnalysis.fromJson(Map<String, dynamic> json) {
    return DraftAnalysis(
      // Tone fields
      tone: json['tone'] as String,
      intensity: json['intensity'] as String?,
      urgencyLevel: json['urgency_level'] as String?,
      intent: json['intent'] as String?,
      contextFlags: json['context_flags'] as Map<String, dynamic>?,
      reasoning: json['reasoning'] as String?,
      
      // Draft fields
      confidenceScore: json['confidence_score'] as int,
      appropriateness: AppropriatenessLevel.fromString(
        json['appropriateness'] as String,
      ),
      suggestions: (json['suggestions'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList() ?? [],
      warnings: (json['warnings'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList() ?? [],
      strengths: (json['strengths'] as List<dynamic>?)
          ?.map((e) => e as String)
          .toList() ?? [],
      
      // NEW: Phase 2 fields
      situationDetection: json['situation_detection'] != null
          ? SituationDetection.fromJson(json['situation_detection'] as Map<String, dynamic>)
          : null,
      suggestedTemplates: (json['suggested_templates'] as List<dynamic>?)
          ?.map((e) => ResponseTemplate.fromJson(e as Map<String, dynamic>))
          .toList(),
    );
  }

  /// Get color based on confidence score
  Color getConfidenceColor() {
    if (confidenceScore >= 90) return Colors.green;
    if (confidenceScore >= 75) return Colors.lightGreen;
    if (confidenceScore >= 60) return Colors.orange;
    return Colors.red;
  }

  /// Get icon based on appropriateness
  IconData getAppropriatenessIcon() {
    switch (appropriateness) {
      case AppropriatenessLevel.excellent:
        return Icons.check_circle;
      case AppropriatenessLevel.good:
        return Icons.thumb_up;
      case AppropriatenessLevel.okay:
        return Icons.info;
      case AppropriatenessLevel.needsWork:
        return Icons.warning;
    }
  }

  /// Get brief status message
  String getStatusMessage() {
    if (confidenceScore >= 90) return 'Ready to send!';
    if (confidenceScore >= 75) return 'Looking good';
    if (confidenceScore >= 60) return 'Could be improved';
    return 'Needs revision';
  }
}

enum AppropriatenessLevel {
  excellent('excellent'),
  good('good'),
  okay('okay'),
  needsWork('needs_work');

  final String value;
  const AppropriatenessLevel(this.value);

  static AppropriatenessLevel fromString(String value) {
    return AppropriatenessLevel.values.firstWhere(
      (e) => e.value == value,
      orElse: () => AppropriatenessLevel.okay,
    );
  }

  String get displayName {
    switch (this) {
      case AppropriatenessLevel.excellent:
        return 'Excellent';
      case AppropriatenessLevel.good:
        return 'Good';
      case AppropriatenessLevel.okay:
        return 'Okay';
      case AppropriatenessLevel.needsWork:
        return 'Needs Work';
    }
  }
}

enum RelationshipType {
  boss('boss', 'Boss/Manager'),
  colleague('colleague', 'Colleague'),
  friend('friend', 'Friend'),
  family('family', 'Family'),
  client('client', 'Client'),
  none('none', 'Not specified');

  final String value;
  final String displayName;
  const RelationshipType(this.value, this.displayName);

  static RelationshipType fromString(String? value) {
    if (value == null) return RelationshipType.none;
    return RelationshipType.values.firstWhere(
      (e) => e.value == value,
      orElse: () => RelationshipType.none,
    );
  }

  IconData get icon {
    switch (this) {
      case RelationshipType.boss:
        return Icons.business;
      case RelationshipType.colleague:
        return Icons.people;
      case RelationshipType.friend:
        return Icons.emoji_people;
      case RelationshipType.family:
        return Icons.family_restroom;
      case RelationshipType.client:
        return Icons.handshake;
      case RelationshipType.none:
        return Icons.help_outline;
    }
  }
}

File: backend/supabase/functions/_shared/prompts/draft_analysis.ts
/**
 * Draft Analysis for Message Confidence Checking
 * Extends enhanced tone analysis for draft/outgoing messages
 */

import {
  ENHANCED_TONE_ANALYSIS_SYSTEM_PROMPT,
  type ToneAnalysisResult,
} from './enhanced-tone-analysis.ts';

import {
  SITUATION_DETECTION_PROMPT,
  detectSituation,
  type SituationType,
  type SituationDetectionResult,
} from './situation-detector.ts';

import {
  ALL_TEMPLATES,
  DECLINING_TEMPLATES,
  BOUNDARY_TEMPLATES,
  INFO_DUMP_TEMPLATES,
  APOLOGIZING_TEMPLATES,
  CLARIFYING_TEMPLATES,
  type ResponseTemplate,
} from '../templates/index.ts';

export interface DraftAnalysisContext {
  draftMessage: string;
  conversationHistory?: string[];
  relationshipType?: 'boss' | 'colleague' | 'friend' | 'family' | 'client' | 'none';
  conversationTone?: string; // From previous tone analysis
  recipientInfo?: {
    name?: string;
    role?: string;
  };
}

export interface DraftAnalysisResult extends ToneAnalysisResult {
  // Extends tone analysis with draft-specific fields
  confidence_score: number; // 0-100
  appropriateness: 'excellent' | 'good' | 'okay' | 'needs_work';
  suggestions: string[];
  warnings: string[];
  strengths: string[];
  
  // NEW: Situation detection and template suggestions
  situation_detection?: SituationDetectionResult;
}

/**
 * Enhanced prompt for draft analysis - extends existing tone analysis with situation detection
 */
export const DRAFT_ANALYSIS_SYSTEM_PROMPT = `${ENHANCED_TONE_ANALYSIS_SYSTEM_PROMPT}

**ADDITIONAL CONTEXT: This is a DRAFT message being composed**

${SITUATION_DETECTION_PROMPT}

Provide additional draft-specific guidance to help the user send with confidence:

1. **Confidence Score** (0-100): How ready is this message to send?
   - 90-100: Excellent - Ready to send, no changes needed
   - 75-89: Good - Minor polish could help, but solid overall
   - 60-74: Okay - Some improvements recommended
   - Below 60: Needs Work - Consider revising before sending

2. **Appropriateness**: Overall assessment (excellent/good/okay/needs_work)

3. **Strengths** (What works well):
   - Highlight positive aspects
   - Encourage what's already good
   - Build confidence

4. **Suggestions** (Maximum 3 specific improvements):
   - Actionable changes
   - Prioritize most impactful
   - Be constructive

5. **Warnings** (Red flags to address):
   - Tone mismatches for relationship context
   - Potentially offensive/ambiguous phrasing
   - Missing key information

**Response Format (JSON):**
{
  "tone": "one of 23 categories",
  "intensity": "one of 5 levels",
  "urgency_level": "one of 4 levels",
  "intent": "3-8 word description",
  "confidence_score": 85,
  "appropriateness": "good",
  "strengths": [
    "Clear and direct communication",
    "Appropriate level of formality"
  ],
  "suggestions": [
    "Consider adding a brief greeting",
    "Could specify deadline more explicitly"
  ],
  "warnings": [],
  "context_flags": {
    "sarcasm_detected": false,
    "tone_indicator_present": false,
    "ambiguous": false
  },
  "situation_detection": {
    "situation_type": "declining",
    "confidence": 0.85,
    "reasoning": "Draft contains decline language",
    "suggested_templates": ["decline_polite", "decline_with_alternative"]
  },
  "reasoning": "Message is professional and clear. Confidence is high..."
}`;

export function generateDraftAnalysisPrompt(context: DraftAnalysisContext): string {
  let prompt = `Analyze this DRAFT message before sending:\n\n`;
  prompt += `**Draft Message:**\n"${context.draftMessage}"\n\n`;
  
  if (context.relationshipType && context.relationshipType !== 'none') {
    prompt += `**Relationship Context:** ${context.relationshipType}\n`;
  }
  
  if (context.conversationTone) {
    prompt += `**Recent Conversation Tone:** ${context.conversationTone}\n`;
  }
  
  if (context.conversationHistory && context.conversationHistory.length > 0) {
    prompt += `**Recent Messages (for context):**\n`;
    context.conversationHistory.slice(-3).forEach((msg, idx) => {
      prompt += `${idx + 1}. "${msg}"\n`;
    });
    prompt += '\n';
  }
  
  if (context.recipientInfo?.name || context.recipientInfo?.role) {
    prompt += `**Recipient:**\n`;
    if (context.recipientInfo.name) prompt += `- Name: ${context.recipientInfo.name}\n`;
    if (context.recipientInfo.role) prompt += `- Role: ${context.recipientInfo.role}\n`;
    prompt += '\n';
  }
  
  prompt += `Provide your analysis in JSON format as specified above.`;
  
  return prompt;
}

export function validateDraftAnalysis(result: any): DraftAnalysisResult {
  // Validate confidence score
  if (typeof result.confidence_score !== 'number' || 
      result.confidence_score < 0 || 
      result.confidence_score > 100) {
    throw new Error('Invalid confidence score');
  }
  
  // Validate appropriateness
  const validAppropriateness = ['excellent', 'good', 'okay', 'needs_work'];
  if (!validAppropriateness.includes(result.appropriateness)) {
    throw new Error('Invalid appropriateness level');
  }
  
  // Validate tone (inherited from ToneAnalysisResult)
  if (!result.tone || typeof result.tone !== 'string') {
    throw new Error('Invalid tone');
  }
  
  return {
    // Tone analysis fields
    tone: result.tone,
    intensity: result.intensity,
    urgency_level: result.urgency_level,
    intent: result.intent,
    confidence_score: result.confidence_score,
    reasoning: result.reasoning,
    context_flags: result.context_flags,
    
    // Draft-specific fields
    appropriateness: result.appropriateness,
    suggestions: result.suggestions || [],
    warnings: result.warnings || [],
    strengths: result.strengths || [],
    
    // NEW: Situation detection
    situation_detection: result.situation_detection,
  };
}

/**
 * Get suggested templates based on detected situation type
 */
export function getSuggestedTemplates(situationType: SituationType): ResponseTemplate[] {
  // Map situation types to template collections
  const templateMap: Record<SituationType, ResponseTemplate[]> = {
    declining: DECLINING_TEMPLATES,
    boundary_setting: BOUNDARY_TEMPLATES,
    info_dumping: INFO_DUMP_TEMPLATES,
    apologizing: APOLOGIZING_TEMPLATES,
    clarifying: CLARIFYING_TEMPLATES,
    casual_chat: [],
    work_professional: [],
    emotional_support: [],
    unknown: [],
  };
  
  return templateMap[situationType] || [];
}

/**
 * Find templates matching specific keywords from the draft
 */
export function findMatchingTemplates(
  draftMessage: string,
  maxResults: number = 5
): ResponseTemplate[] {
  const messageLower = draftMessage.toLowerCase();
  const matches: Array<{ template: ResponseTemplate; score: number }> = [];
  
  for (const template of ALL_TEMPLATES) {
    let score = 0;
    
    // Check context keywords
    for (const keyword of template.context) {
      if (messageLower.includes(keyword.toLowerCase())) {
        score += 1;
      }
    }
    
    // Boost score for neurodivergent-friendly templates
    if (template.neurodivergent_friendly) {
      score += 0.5;
    }
    
    if (score > 0) {
      matches.push({ template, score });
    }
  }
  
  // Sort by score and return top results
  matches.sort((a, b) => b.score - a.score);
  return matches.slice(0, maxResults).map(m => m.template);
}

